crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
             .member(db, &attr.id)
             .unwrap_with_diagnostic(|lookup_error| match lookup_error {
                 LookupError::Unbound => {
+                    let bound_on_instance = match value_type {
+                    Type::ClassLiteral(ClassLiteralType { class }) => {
+                        !class.instance_member(self.db(), attr).0.is_unbound()
+                    }
+                    Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                        match subclass_of.subclass_of() {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
+                            }
+                            ClassBase::Dynamic(_) => unreachable!("For any symbol, lookup on a dynamic type will return a bound symbol of the same type"),
```

Comment:
a couple of nits:

- Lookup on a dynamic type won't always return a bound symbol -- e.g. `tuple[Any, Any]` is a dynamic type, but looking up the attribute "foo" on `tuple[Any, Any]` won't return a bound symbol. This is specifically true for dynamic `SubclassOf` types, but not true in general for _all_ dynamic types.
- It won't necessarily return a bound symbol of the same _type_. That might be what we have right now, but in the long term, for example, `str & Any` might be a better answer for the type of the `__module__` attribute on an object with type `type[Any]`.

```suggestion
                            ClassBase::Dynamic(_) => unreachable!("Attribute lookup on a dynamic `SubclassOf` type should always return a bound symbol"),
```